src/google_civic/core.clj

```diff
@@ -1,39 +1,106 @@
 (ns google-civic.core
   (:require [clj-http.client :as client]
             [cheshire.core :as json]
-            [slingshot.slingshot :refer [try+]]))
+            [clojure.set :as set]))
 
 (def api-root "https://www.googleapis.com/civicinfo/v2")
 
+(def temporary-rate-limit-reasons
+  #{"concurrentLimitExceeded" "rateLimitExceeded"
+    "servingLimitExceeded" "userRateLimitExceeded"})
+
+(defn temporary-rate-limit?
+  "True if the errors in the response only include errors from the
+  temporary-rate-limit-reasons set and the code is 403."
+  [response body]
+  (and (= 403 (:status response))
+       (let [errors (get-in body [:error :errors])
+             reasons (->> errors (map :reason) set)]
+         (and (seq reasons)
+              (set/superset? temporary-rate-limit-reasons
+                             reasons)))))
+
+(defn can-retry?
+  "We can retry if the response code indicates a server error, or
+   if the error is deemed a temporary rate limit."
+  [response body]
+  (cond
+    (client/server-error? response) true
+    (temporary-rate-limit? response body) true
+    :else
+    false))
+
 (defn api-url [endpoint]
   (str api-root endpoint))
 
 (defn api-response [api-key endpoint & [query-params]]
   (let [url (api-url endpoint)]
     (client/get url {:accept :json
+                     :throw-exceptions false
                      :query-params (merge {:key api-key} query-params)})))
 
 (defn api-req
-  ([api-key endpoint & [query-params]]
-   (try+
-     (let [response (api-response api-key endpoint query-params)]
-       (json/parse-string (:body response) true))
-     (catch Object response
-       (merge response {:body (json/parse-string (:body response) true)})))))
-
-(defn elections [api-key]
-  (let [response (api-req api-key "/elections")]
-    (if-let [elections (:elections response)]
-      elections
-      (:body response))))
-
-(defn voter-info [api-key addr & [election-id]]
-  (let [query-params {:address addr}
-        query-params (if election-id
-                       (merge query-params {:electionId election-id})
-                       query-params)
-        response (api-req api-key (str "/voterinfo") query-params)
-        voter-info (dissoc response :kind)]
-    (if (empty? voter-info)
-      (:body response)
-      voter-info)))
+  ([api-key endpoint]
+   (api-req api-key endpoint nil))
+  ([api-key endpoint query-params]
+   (api-req api-key endpoint query-params 0 nil))
+  ([api-key endpoint query-params retry-limit backoff-fn]
+   (loop [response (api-response api-key endpoint query-params)
```

Comment:
It seems to do a lot of what you are doing here wrapped up in a well tested library (it's been in production across hundreds of AWS nodes for since 2011). Again, I value working code over using the library and the main extra thing I think it provides is plugable backoff strategies, however I'm not sure if we would use that.